README
NAME
    Class::Accessor::Installer - Install an accessor subroutine

VERSION
    version 1.100820

SYNOPSIS
        package Class::Accessor::Foo;

        use base 'Class::Accessor::Installer';

        sub mk_foo_accessors {
            my ($self, @fields) = @_;
            my $class = ref $self || $self;

            for my $field (@fields) {
                $self->install_accessor(
                    sub     => "${field}_foo",
                    code    => sub { rand() },
                    purpose => 'Does this, that and the other',
                    example => [
                        "my \$result = $class->${field}_foo(\$value)",
                        "my \$result = $class->${field}_foo(\$value, \$value2)",
                    ]
                );
            }
        }

DESCRIPTION
    This mixin class provides a method that will install a coderef. There
    are other modules that do this, but this one is a bit more specific to
    the needs of Class::Accessor::Complex and friends.

    It is intended as a mixin, that is, your accessor-generating class
    should inherit from this class.

METHODS
  install_accessor
    Takes as arguments a named hash. The following keys are recognized:

    "pkg"
        The package into which to install the subroutine. If this argument
        is omitted, it will inspect $self to determine the package.
        Class::Accessor::* accessor generators are typically used like this:

            __PACKAGE__
                ->mk_new
                ->mk_array_accessors(qw(foo bar));

        Therefore "install_accessor()" can determine the right package into
        which to install the subroutine.

    "name"
        The name or names to use for the subroutine. You can either pass a
        single string or a reference to an array of strings. Each string is
        interpreted as a subroutine name inside the given package, and the
        code reference is installed into the appropriate typeglob.

        Why would you want to install a subroutine in more than one place
        inside your package? For example, Class::Accessor::Complex often
        creates aliases so the user can choose the version of the name that
        reads more naturally.

        An example of this usage would be:

                $self->install_accessor(
                    name => [ "clear_${field}", "${field}_clear" ],
                    code => sub { ... }
                );

    "code"
        This is the code reference that should be installed.

    "purpose"
        A string describing the generated method. This information can be
        used by Pod::Generated to automatically generate pod documentation
        during "make" time.

    "example"
        One or more examples of using the method. These will also be used in
        the generated documentation. The value can be a string or an
        reference to an array of strings.

    The installed subroutine is named using Sub::Name, so it shows up with a
    meaningful name in stack traces (instead of as "__ANON__"). However, the
    inside the debugger, the subroutine still shows up as "__ANON__". You
    might therefore want to use the following lines at the beginning of your
    subroutine:

            $self->install_accessor(
                name => $field,
                code => sub {
                    local $DB::sub = local *__ANON__ = "${class}::${field}"
                        if defined &DB::DB && !$Devel::DProf::VERSION;
                    ...
            );

    Now the subroutine will be named both in a stack trace and inside the
    debugger.

INSTALLATION
    See perlmodinstall for information and options on installing Perl
    modules.

BUGS AND LIMITATIONS
    No bugs have been reported.

    Please report any bugs or feature requests through the web interface at
    <http://rt.cpan.org/Public/Dist/Display.html?Name=Class-Accessor-Install
    er>.

AVAILABILITY
    The latest version of this module is available from the Comprehensive
    Perl Archive Network (CPAN). Visit <http://www.perl.com/CPAN/> to find a
    CPAN site near you, or see
    <http://search.cpan.org/dist/Class-Accessor-Installer/>.

    The development version lives at
    <http://github.com/hanekomu/Class-Accessor-Installer/>. Instead of
    sending patches, please fork this project using the standard git and
    github infrastructure.

AUTHORS
      Marcel Gruenauer <marcel@cpan.org>
      Florian Helmberger <florian@cpan.org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2007 by Marcel Gruenauer.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for Perl extension Class-Accessor-Installer

1.100820  2010-03-23 18:13:50 Europe/Vienna
    - converted distribution to Dist::Zilla-style

0.09 Wed 2010.02.10 14:34:59 CET (Marcel Gruenauer <marcel@cpan.org>)
     - removed optional but - after the Vim::Tag rewrite unnecessary - Vim::Tag
       dependency, thereby also removing the circular dependency of Vim::Tag ->
       Class::Accessor::Constructor -> Class::Accessor::Complex ->
       Class::Accessor::Installer.

0.08 Sun 2010.02.07 22:20:43 CET (Marcel Gruenauer <marcel@cpan.org>)
     - removed test dependency on Class::Accessor - works fine without it

0.07 Mon 2010.02.01 22:38:08 CET (Marcel Gruenauer <marcel@cpan.org>)
     - convert to new-style PTAGS handling (see Vim::Tag)

0.06 Sun 2009.12.06 23:41:53 CET (Marcel Gruenauer <marcel@cpan.org>)
     - Vim::Tag is now an optional feature
     - reduced dependencies using author tests
     - better dist engineering (INSTALL, README, README.mkdn etc.)

0.05  Sat Nov  1 22:01:44 CET 2008 (Marcel Gruenauer <marcel@cpan.org>)
     - Now uses Vim::Tag

0.04  Thu Oct 30 16:23:28 CET 2008 (Marcel Gruenauer <marcel@cpan.org>)
     - Generate ptags if $::PTAGS is set (Florian Helmberger)

0.03  Thu Jul 25 15:27:43 CEST 2008 (Marcel Gruenauer <marcel@cpan.org>)
     - Changed dist style and Changes back to standard. People didn't like it -
       the nail that sticks out gets hammered down.
     - Added standard test files; this will also help with CPANTS.

0.02  Thu, 29 May 2008 12:01:28 -0000 (Marcel Gruenauer <marcel@cpan.org>)
     - Moved to YAML Changes file
     - Added auto_install feature 'Generated documentation' that requires
       Pod::Generated
     - .shipit: added Twitter step
     - Makefile.PL: added auto_install() and process_templates()
     - lib/*: converted to template
     - updated MANIFEST
     - add documentation on the 'purpose' and 'example' keys taken by
       install_accessor()
     - updated t/perlcriticrc
     - Changes: fixed global.name
     - .shipit: fixed svk.tagpattern
     - tags: NEWFEATURE

0.01  Tue, 23 Oct 2007 21:17:37 +0200 (Marcel Gruenauer <marcel@cpan.org>)
     - original version


lib/Class/Accessor/Installer.pm
use 5.006;
use warnings;
use strict;

package Class::Accessor::Installer;
our $VERSION = '1.100820';
# ABSTRACT: Install an accessor subroutine
use Sub::Name;
use UNIVERSAL::require;

sub install_accessor {
    my ($self, %args) = @_;
    my ($pkg, $name, $code) = @args{qw/pkg name code/};
    unless (defined $pkg) {
        $pkg = ref $self || $self;
    }
    $name = [$name] unless ref $name eq 'ARRAY';
    my @caller;
    if ($::PTAGS) {
        my $level = 1;
        do { @caller = caller($level++) }
          while $caller[0] =~ /^Class(::\w+)*::Accessor::/o;
    }
    for my $sub (@$name) {
        no strict 'refs';
        $::PTAGS && $::PTAGS->add_tag($sub, $caller[1], $caller[2]);
        *{"${pkg}::${sub}"} = subname "${pkg}::${sub}" => $code;
        for my $doc_type (qw(purpose example)) {
            next unless defined $args{$doc_type};

            # don't use() it - this installer should still work if we don't
            # have Pod::Generated
            Pod::Generated->require;
            next if $@;
            my $spec = $args{$doc_type};
            $spec = [$spec] unless ref $spec eq 'ARRAY';
            for my $doc_el (@$spec) {
                $doc_el =~ s/^\s*|\s*$//sg;
                Pod::Generated::add_doc($pkg, 'CODE', $sub, $doc_type, $doc_el);
            }
        }
    }
}
1;


__END__
=pod

=head1 NAME

Class::Accessor::Installer - Install an accessor subroutine

=head1 VERSION

version 1.100820

=head1 SYNOPSIS

    package Class::Accessor::Foo;

    use base 'Class::Accessor::Installer';

    sub mk_foo_accessors {
        my ($self, @fields) = @_;
        my $class = ref $self || $self;

        for my $field (@fields) {
            $self->install_accessor(
                sub     => "${field}_foo",
                code    => sub { rand() },
                purpose => 'Does this, that and the other',
                example => [
                    "my \$result = $class->${field}_foo(\$value)",
                    "my \$result = $class->${field}_foo(\$value, \$value2)",
                ]
            );
        }
    }

=head1 DESCRIPTION

This mixin class provides a method that will install a coderef. There are
other modules that do this, but this one is a bit more specific to the needs
of L<Class::Accessor::Complex> and friends.

It is intended as a mixin, that is, your accessor-generating class should
inherit from this class.

=head1 METHODS

=head2 install_accessor

Takes as arguments a named hash. The following keys are recognized:

=over 4

=item C<pkg>

The package into which to install the subroutine. If this argument is omitted,
it will inspect C<$self> to determine the package. Class::Accessor::*
accessor generators are typically used like this:

    __PACKAGE__
        ->mk_new
        ->mk_array_accessors(qw(foo bar));

Therefore C<install_accessor()> can determine the right package into which to
install the subroutine.

=item C<name>

The name or names to use for the subroutine. You can either pass a single
string or a reference to an array of strings. Each string is interpreted as a
subroutine name inside the given package, and the code reference is installed
into the appropriate typeglob.

Why would you want to install a subroutine in more than one place inside your
package? For example, L<Class::Accessor::Complex> often creates aliases so the
user can choose the version of the name that reads more naturally.

An example of this usage would be:

        $self->install_accessor(
            name => [ "clear_${field}", "${field}_clear" ],
            code => sub { ... }
        );

=item C<code>

This is the code reference that should be installed.

=item C<purpose>

A string describing the generated method. This information can be used by
L<Pod::Generated> to automatically generate pod documentation during C<make>
time.

=item C<example>

One or more examples of using the method. These will also be used in the
generated documentation. The value can be a string or an reference to an array
of strings.

=back

The installed subroutine is named using L<Sub::Name>, so it shows up with a
meaningful name in stack traces (instead of as C<__ANON__>). However, the
inside the debugger, the subroutine still shows up as C<__ANON__>. You might
therefore want to use the following lines at the beginning of your subroutine:

        $self->install_accessor(
            name => $field,
            code => sub {
                local $DB::sub = local *__ANON__ = "${class}::${field}"
                    if defined &DB::DB && !$Devel::DProf::VERSION;
                ...
        );

Now the subroutine will be named both in a stack trace and inside the
debugger.

=head1 INSTALLATION

See perlmodinstall for information and options on installing Perl modules.

=head1 BUGS AND LIMITATIONS

No bugs have been reported.

Please report any bugs or feature requests through the web interface at
L<http://rt.cpan.org/Public/Dist/Display.html?Name=Class-Accessor-Installer>.

=head1 AVAILABILITY

The latest version of this module is available from the Comprehensive Perl
Archive Network (CPAN). Visit L<http://www.perl.com/CPAN/> to find a CPAN
site near you, or see
L<http://search.cpan.org/dist/Class-Accessor-Installer/>.

The development version lives at
L<http://github.com/hanekomu/Class-Accessor-Installer/>.
Instead of sending patches, please fork this project using the standard git
and github infrastructure.

=head1 AUTHORS

  Marcel Gruenauer <marcel@cpan.org>
  Florian Helmberger <florian@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2007 by Marcel Gruenauer.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



Makefile.PL

use strict;
use warnings;

 BEGIN { require 5.006; } 

use ExtUtils::MakeMaker 6.11;



my %WriteMakefileArgs = (
                       'test' => {
                                   'TESTS' => 't/*.t'
                                 },
                       'NAME' => 'Class::Accessor::Installer',
                       'DISTNAME' => 'Class-Accessor-Installer',
                       'CONFIGURE_REQUIRES' => {
                                                 'ExtUtils::MakeMaker' => '6.11'
                                               },
                       'AUTHOR' => 'Marcel Gruenauer <marcel@cpan.org>, Florian Helmberger <florian@cpan.org>',
                       'BUILD_REQUIRES' => {},
                       'ABSTRACT' => 'Install an accessor subroutine',
                       'EXE_FILES' => [],
                       'VERSION' => '1.100820',
                       'PREREQ_PM' => {
                                        'Test::More' => '0.88',
                                        'Scalar::Util' => '0',
                                        'English' => '0',
                                        'File::Find' => '0',
                                        'File::Temp' => '0',
                                        'Sub::Name' => '0',
                                        'UNIVERSAL::require' => '0',
                                        'Carp' => '0'
                                      },
                       'LICENSE' => 'perl'
                     );


delete $WriteMakefileArgs{LICENSE}
  unless eval { ExtUtils::MakeMaker->VERSION(6.31) };

WriteMakefile(%WriteMakefileArgs);



